### PR TITLE
Pre-upload album art for tracks in playing queue

### LINKED ISSUE #94: Pre-Upload Assets from current playlist
Option to upload artwork from the next x titles in the playlist so they are available to display when the title actually starts playing.

### PR DESCRIPTION
**Pull request by VelocityRa**
Hardcoded to pre-uploading the 5 tracks after the current one.

Resolves #94.

### CODE REVIEW

File: DiscordBee.cs [hunk 1]
```diff
@@ -128,6 +128,8 @@ public void ReceiveNotification(string sourceFileUrl, NotificationType type)
         case NotificationType.TrackChanged:
         case NotificationType.PlayStateChanged:
           UpdateDiscordPresence(_mbApiInterface.Player_GetPlayState());
+          if (type == NotificationType.TrackChanged)
```
Comment (sll552) [inline DiscordBee.cs:131]: Why is this needed? isn't that covered by `PlayingTracksChanged` already?

File: DiscordBee.cs [hunk 2]
```diff
@@ -136,6 +138,27 @@ public void ReceiveNotification(string sourceFileUrl, NotificationType type)
...
+      }
+    }
+
+    private void UploadQueuedTracks()
+    {
+      for (int i = 0; i < 4; i++)
+      {
+        int nextPlayingIndex = _mbApiInterface.NowPlayingList_GetCurrentIndex();
+        string fileUrl = _mbApiInterface.NowPlayingList_GetListFileUrl(nextPlayingIndex + 1 + i);
+        if (fileUrl == null || fileUrl == "")
```
Comment (sll552) [inline DiscordBee.cs:153]: There is https://docs.microsoft.com/en-us/dotnet/api/system.string.isnullorempty?view=net-6.0
Also please always add braces to ifs even for oneliners

File: DiscordTools/DiscordClient.cs [hunk 3]
```diff
@@ -114,6 +114,34 @@ public void Close()
...
+          {
+            assetId = AssetManager.ASSET_LOGO;
+            _ = _assetManager.UploadAsset(artworkData).ContinueWith(
+              t =>
+              {
+                var id = t.Result;
+                if (id == null)
+                { // this most likely means another task is already uploading this asset.
+                return;
+                }
```
Comment (sll552) [inline DiscordTools/DiscordClient.cs:137]: Why did you remove the presence update?
Maybe I am missing something here but if the artwork matches the current track (can be checked using `_currentArtworkHash`) it should update the presence if it finished uploading.

**Review by sll552 — CHANGES_REQUESTED** [review]

File: DiscordBee.cs (hunk 1 above)
Comment (VelocityRa) [reply DiscordBee.cs:131]: Nope, that event is not fired when the current track ends and the next one begins.

File: DiscordTools/DiscordClient.cs (hunk 3 above)
Comment (VelocityRa) [reply DiscordTools/DiscordClient.cs:137]: Nah you're right, it's barely any work, and we can deduplicate the code.

Comment (VelocityRa) [reply DiscordTools/DiscordClient.cs:137]: Wait no, I remembered why I didn't do this:
https://github.com/sll552/DiscordBee/blob/master/DiscordTools/DiscordClient.cs#L119
`RichPresence` is cloned at call time. Unless I'm mistaken, if we se this in the `ContinueWith` with just `LargeImageKey` changed, it would set the correct image, but everything else would be wrong (whatever it was when the image was requested to be preloaded).

Comment (VelocityRa) [reply DiscordTools/DiscordClient.cs:137]: I'm not sure why but this doesn't seem to be a problem actually. I'm not sure why but then again I didn't test that much.

Comment (VelocityRa) [reply DiscordTools/DiscordClient.cs:137]: Did more testing, seems fine, can probably merge. Pushing now.

Comment (VelocityRa) [reply DiscordTools/DiscordClient.cs]: Ohh I see why it's not an issue. `_discordPresence` is a class member 🙃 

Imo it's a bit confusing that the underscore naming convention is used both for class members, but also for locals in some places.

Comment (VelocityRa) [reply DiscordTools/DiscordClient.cs]: This was a bit annoying to do right when I tried so I stopped working on it, idk if/when I'll pick it up.

Comment (sll552) [reply DiscordTools/DiscordClient.cs]: Please test my take on this, seemed working for me but with the caching being so inconsistent it might as well be by pure luck.

**Review by sll552 — APPROVED** [review]
